src/shared/awsFiletypes.ts

```diff
@@ -0,0 +1,114 @@
+ */
+export function langidToAwsFiletype(langId: string): telemetry.AwsFiletype {
+    switch (langId) {
+        case constants.ssmJson:
+        case constants.ssmYaml:
+            return 'ssmDocument'
+        case aslFormats.JSON_ASL:
+        case aslFormats.YAML_ASL:
+            return 'stepfunctionsAsl'
+        default:
+            return 'other'
+    }
+}
+
+/** Returns true if file `f` is somewhere in `~/.aws`. */
+export function isAwsConfig(f: string): boolean {
+    const awsDir = path.join(sysutil.SystemUtilities.getHomeDirectory(), '.aws')
+    if (fsutil.isInDirectory(awsDir, f)) {
+        return true
+    }
+    return false
+}
+
+export function isAwsFiletype(doc: vscode.TextDocument): boolean | undefined {
+    if (awsFiletypeLangIds.awsOwned.includes(doc.languageId)) {
+        return true
+    }
+    if (isAwsConfig(doc.fileName)) {
+        return true
+    }
+    if (awsFiletypeLangIds.ambiguous.includes(doc.languageId)) {
+        return undefined // Maybe
+    }
+    return false
+}
+
+export function activate(): void {
+    globals.context.subscriptions.push(
+        // TODO: onDidChangeTextDocument ?
+        vscode.workspace.onDidOpenTextDocument(async (doc: vscode.TextDocument) => {
+            const isAwsFileExt = isAwsFiletype(doc)
+            const isSchemaHandled = globals.schemaService.isMapped(doc.uri)
+            const isCfnTemplate = !!globals.templateRegistry.registeredItems.find(
+                t => pathutil.normalize(t.path) === pathutil.normalize(doc.fileName)
+            )
+            if (!isAwsFileExt && !isSchemaHandled && !isCfnTemplate) {
+                return
+            }
+
+            let fileExt: string | undefined = path.extname(doc.fileName)
+            fileExt = fileExt ? fileExt : undefined // Telemetry client will fail on empty string.
+
+            // TODO: ask templateRegistry if this is SAM or CFN.
+            // TODO: ask schemaService for the precise filetype.
+            let telemKind = isAwsConfig(doc.fileName) ? 'awsCredentials' : langidToAwsFiletype(doc.languageId)
+            if (telemKind === 'other') {
+                telemKind = isCfnTemplate ? 'cloudformationSam' : isSchemaHandled ? 'cloudformation' : 'other'
+            }
+
+            // HACK: for "~/.aws/foo" vscode sometimes _only_ emits "~/.aws/foo.git".
+            if (telemKind === 'awsCredentials' && fileExt === '.git') {
+                fileExt = undefined
+            }
+
+            if (await isSameMetricPending(telemKind, fileExt)) {
+                return // Avoid redundant/duplicate metrics.
+            }
+
+            telemetry.recordFileEditAwsFile({
+                awsFiletype: telemKind,
+                passive: true,
+                result: 'Succeeded',
+                filenameExt: fileExt,
+            })
+        }, undefined)
+    )
+}
+
+async function isSameMetricPending(filetype: string, fileExt: string | undefined): Promise<boolean> {
```

Comment:
The metric is now directly dependent on how the telemetry subsystem works. It would be better to debounce metric emission with a constant time. Easier to reason about.

Reply:
https://github.com/aws/aws-toolkit-vscode/issues/2672